text/0000-dyn-assoc.md

```diff
@@ -0,0 +1,283 @@
+
+## Associated constants
+It's not any harder for associated constants, since these can be simply stored in the vtable along with the methods and associated functions. The only reason why this might become an implementation issue is that the vtable lookup code can no longer assume that all elements have the same size alignment. It's of high doubt that this assumption is ever made anywhere inside the compiler, though.
+
+## Mentions of the `Self` type in methods and associated functions
+As [explained back in the guide section][mentions-of-the-implementing-type-in-assoc-fn-guide], `Self` as a concrete type is still not allowed in traits if these wish to stay object-safe. Trait objects of a trait in the declaration of that exact same trait are perfectly fine, and even [do compile in present Rust](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=45c35ab9274c22b323a9016796c635a5):
+```rust
+trait Thing {
+    fn make_a_friend(&self) -> Box<dyn Thing>;
+}
+struct SimpleThing;
+impl Thing for SimpleThing {
+    fn make_a_friend(&self) -> Box<dyn Thing> {
+        Box::from(SimpleThing)
+    }
+}
+struct BetterThing;
+impl Thing for BetterThing {
+    fn make_a_friend(&self) -> Box<dyn Thing> {
+        Box::from(BetterThing)
+    }
+}
+
+fn main() {
+    let sithing: Box<dyn Thing> = Box::from(SimpleThing);
+    let bething: Box<dyn Thing> = Box::from(BetterThing);
+    let sithing2 = sithing.make_a_friend();
+    let bething2 = sithing.make_a_friend();
+}
+```
+Supporting the actual `Self` type [might still be introduced in a later RFC][future-possibilities].
+
+## Generic contexts
+As [described][generic-contexts] in the guide-level explanation, we can put trait objects themselves in the generics to get associated members to work in functions which do not take the trait object as an argument and are not dynamically dispatched from the side which has the trait object. How does this work?
+
+For functions with generic arguments, the vtable (not the trait object, only its vtable) is passed as an invisible argument on the stack/registers, which is present only in the morphomization of the function which takes a trait object in its generics. If a function has more than one `?Sized` generic argument, the morphomization (generated at compile time) which requires multiple trait objects should take multiple invisible vtable arguments. The arguments are automatically filled out when the callee puts  the trait object(s) into the angle brackets (if they don't provide all the trait objects required and fill some with `dyn Trait`, an error is produced as described back in the guide-level section).
```

Comment:
Where is this vtable supposed to come from? You don't have a value, so you cannot pass a vtable. In the example I gave in the comments, no implementation of the trait even exists at all. Also, for a generic function `foo::<T: Trait + ?Sized>()`, the function `foo<dyn Trait>` has type `fn()`, there is no way to pass a hidden argument.

And btw, you cannot simply generate errors when someone passes `dyn Trait` as a type parameter. This is a breaking change.

In my opinion, this entire proposal is contradictory to how Rust generics and traits work and could not be implemented without breaking changes.

Reply:
> Where is this vtable supposed to come from? You don't have a value, so you cannot pass a vtable.

Have you read the guide-level explanation? It clearly states that the vtable comes from passing the trait object (as an rvalue) between the angle brackets.

> In the example I gave in the comments, no implementation of the trait even exists at all. Also, for a generic function `foo::<T: Trait + ?Sized>()`, the function `foo<dyn Trait>` has type `fn()`, there is no way to pass a hidden argument.

There is, the language ABI is not considered stable. The compiler can insert all sorts of invisible implicit arguments as long as it keeps track of what it inserted. This doesn't even break function pointers, since you'd need the generic parameters with trait bounds in the function pointer which would the invisible argument to the pointer type.

> And btw, you cannot simply generate errors when someone passes `dyn Trait` as a type parameter. This is a breaking change.

The only case when `dyn Trait` is invalid as a type parameter is if the function/associated function of the type accepting said parameter uses the associated members of the type parameter. Since currently it's **impossible to access the associated members in the first place**, this doesn't break anything.

Reply:
> It clearly states that the vtable comes from passing the trait object (as an rvalue) between the angle brackets.

You don't see the problem. There is no trait object, you have a type. The type is `dyn Tr` for some trait `Tr`. Any associated item (be it a constant or a function) is a property of that type. An expression like `<T as Tr>::associated_item` is legal Rust, and so is `<dyn Tr as Tr>::associated_item`. Changing that is a breaking change.

> There is, the language ABI is not considered stable.

This is not an ABI issue, this is a language issue. If we have a function `foo<T>()`, for any concrete type `T1` (and `dyn Tr` is a concrete type), the function `foo::<T1>` has type `fn()`. That is part of the language and existing code relies on that. What you suggest here is breaking existing code.

> The only case when dyn Trait is invalid as a type parameter is if the function/associated function of the type accepting said parameter uses the associated members of the type parameter.

Again, you don't seem to understand the issue at all. This is valid Rust code today:

```rust
// Add any object-safe trait here:
trait Tr1 {}

fn with_tr1<T: Tr1 + ?Sized>() {}

fn example() {
    with_tr1<dyn Tr1>();
}
```

And this is valid Rust code today:

```rust
trait Tr2 {
    const A: usize;
}

fn with_tr2<T: Tr2 + ?Size>() {
    let _ = <T as Tr2>::A;
}

// Invalid today, because Tr2 is not object-safe
//fn example2() {
//    with_tr2<dyn Tr2>();
//}
```
Your proposal would make `Tr2` object safe, making the above commeted out function legal. Your RFC does not explain at all what the semantics of this would be.

Reply:
> An expression like `<T as Tr>::associated_item` is legal Rust, and so is `<dyn Tr as Tr>::associated_item`. Changing that is a breaking change.

`<dyn Tr as Tr>::associated_item` never was, and will not be valid. As I said, the compiler produces a compile-time error because there's no vtable, **just like it did before the RFC**.

> If we have a function `foo<T>()`, for any concrete type `T1` (and `dyn Tr` is a concrete type), the function `foo::<T1>` has type `fn()`. That is part of the language and existing code relies on that. What you suggest here is breaking existing code.

The language-level signature, yes. `fn()` means that it takes no **explicit** arguments, i.e. is always called without them. To call the `dyn Trait` morphomization of the function, however, you need to provide the vtable in the angle brackets instead of putting `dyn Trait` there, and, once again, this does not break anything because not a single function before has called a function querying the associated members of a type with `dyn Trait` as a template parameter because these traits were never object-safe. This means that the change is invisible at source level, since no one ever did that kind of thing before. Once again, the place where the error actually happens is the calling function, if it calls a function which uses the associated members of a trait with `dyn Trait` as the template parameter, which was illegal before this RFC.

The reason `fn()` causes no problems whatsoever is because the amount of **arguments** it takes is **zero**, even with the trait object vtable angle bracket trick this RFC proposes. The place where the vtable goes is invisible to **everything and everyone** at language level, because the generic parameter vtable argument **does not exist at language level**, which is why it's called invisible in the first place, and is part of the calling convention, which is ABI.

> Your proposal would make `Tr2` object safe, making the above commeted out function legal. 

It **does not** make the commented out function legal. It doesn't supply a vtable, which is why it's illegal.

> Your RFC does not explain at all what the semantics of this would be.

It does, there's a separate paragraph about the fact that you need a vtable if you're calling a function which accesses associated members of a trait with `dyn Trait` as a template parameter, which, once again, was never allowed.

Reply:
> It does not make the commented out function legal.

Yes, it does.

1) The generic function `fn with_foo<T: Foo + ?Sized>()` is legal. It can be called for any *type*`T1` that implements `Foo`. Also, for any such *type* `T1`, the expression `with_foo::<T1>` has type `fn()`. (This is working in stable Rust and must not break.)
2) In the body of said function, you can access any associated item of the trait `Foo`, including associated constants and functions without a receiver. (This is working in stable Rust and must not break.)
3) If `Foo` is object-safe `dyn Foo` is a valid type that implements `Foo`. (This is working in stable Rust and must not break.)

Combine these rules (which are guaranteed by Rust's stability guarantees) and the result is what I described in the above comments.

Reply:
> `<dyn Tr as Tr>::associated_item` never was, and will not be valid. As I said, the compiler produces a compile-time error because there's no vtable, **just like it did before the RFC**.

This is simply wrong.

The expression `<T as Tr>::item` is valid for any trait `Tr`, any type `T` that implements `Tr` and any item `item` of the trait `Tr` (in stable Rust, today, see [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0b38e06b0c61b50c80c21b2e33dbe839) for an example). If `Tr` is object-safe, `dyn Tr` is a type that implements `Tr`. Thus, for any item `item` of `Tr`, `<dyn Tr as Tr>::item` is valid.

I feel like I'm repeating myself (which I am probably), but you don't seem to understand that these things work **at the type level** only, in the **absence of values**. And in generic contexts, there is no way to know if a type `T` is a struct/enum type or a trait object type, it's simply a type.